docs/articles/multi-attribute-semantic-search.md

```diff
@@ -0,0 +1,483 @@
+
+1. embed objects as undifferentiated single vectors (*single-vector embedding*), search attributes separately, then weight and combine your results
+2. embed attributes separately (*multi-vector embedding*), weight and combine attribute vectors, then perform a single search
+
+Below, we'll walk through an implementation (using a role-play game scenario) of the single-vector embedding approach, which, despite its intuitive appeal, performs poorly even in simple cases. Then, we'll implement the multi-vector embedding approach using the Superlinked framework, which performs far better. Finally, we'll explain why.
+
+These simple implementations will illustrate how to create more powerful and flexible search systems, ones that can handle complex, multi-faceted queries with ease.
+
+## Setting
+
+It's game night, your friends are sitting eagerly around the game table, ready to become characters and embark on an epic quest of Dungeons & Dragons (D&D). Tonight, you're Dungeon Master (storyteller and guide), crafter of thrilling encounters that challenge and excite your players. 
+
+Your trusty Monster Manual contains thousands of creatures, but finding the perfect monster for each situation can be overwhelming. The ideal foe needs to match the setting, difficulty, and narrative of the moment – all while surprising and delighting your players.
+
+What if we could create a tool that instantly finds the most suitable monsters for any scenario? A tool that considers multiple factors simultaneously, ensuring each encounter is as immersive and exciting as possible?
+
+Let's embark on our own quest to build this ultimate monster-finding system, using the power of multi-attribute vector search!
+
+## Dataset
+
+To demonstrate our search techniques, we'll use a small synthetic dataset generated by a Large Language Model (LLM). Here's the prompt used to create our monster database:
+
+```
+Generate two JSON lists: 'monsters' and 'queries'.
+
+1. 'monsters' list: Create 20 unique monsters with the following properties:
+   - name: A distinctive name
+   - look: Brief description of appearance (2-3 sentences)
+   - habitat: Where the monster lives (2-3 sentences)
+   - behavior: How the monster acts (2-3 sentences)
+
+   Ensure some monsters share similar features while remaining distinct.
+
+2. 'queries' list: Create 5 queries to search for monsters:
+   - Each query should be in the format: {look: "...", habitat: "...", behavior: "..."}
+   - Use simple, brief descriptions (1-3 words per field)
+   - Make queries somewhat general to match multiple monsters
+
+Output format:
+{
+  "monsters": [
+    {"name": "...", "look": "...", "habitat": "...", "behavior": "..."},
+    ...
+  ],
+  "queries": [
+    {"look": "...", "habitat": "...", "behavior": "..."},
+    ...
+  ]
+}
+```
+
+Let's examine a sample of our generated dataset. Note that your results may differ due to the non-deterministic nature of LLM generation.
+
```

Comment:
why?

Reply:
@robertdhayanturner 
This is because of internal implementation of OpenAI that is not shared with anyone and can be changed without any notice. There are some [suggestions by OpenAI itself](https://platform.openai.com/docs/advanced-usage/reproducible-outputs) on how to reduce (not completely avoid) randomness in LLM outputs. And the most important argument comes for the experiments :) People try to use some work-arounds but there is no guarantee of deterministic outputs.